### PR TITLE
Makefile: restore msan support

### PR DESCRIPTION
**Pull request by tamird**
Also fixes an issue where caches were not properly invalidated when
switching between jemalloc and stdmalloc builds.

### CODE REVIEW

**Review by tbg — COMMENTED** [review]
Too science dog to really stamp this, but looks good to me.

File: Makefile [hunk 1]
```diff
@@ -56,40 +56,96 @@ TYPE :=
 # toolchain.
 LINKFLAGS ?=
 
+BUILD_TYPE := development
 ifeq ($(TYPE),)
-override LINKFLAGS += -X github.com/cockroachdb/cockroach/pkg/build.typ=development
+else ifeq ($(TYPE),msan)
+NATIVE_SUFFIX := _msan
+override GOFLAGS += -msan
+# NB: using jemalloc with msan causes segfaults. Why?
```
Comment (tschottdorf) [inline Makefile]: maybe file this instead? Maybe we can lure someone in to discuss it.

Comment (tamird) [inline Makefile]: I think we'd need the minimal repro to file it; for now I was discouraged by https://github.com/jemalloc/jemalloc/issues/558#issuecomment-272982746 from producing that repro, but I may revisit.

Interestingly, it doesn't look like anyone has tried using the two together, and even firefox's msan support turns off jemalloc (search for `https://developer.mozilla.org/en-US/docs/Memory_Sanitizer` in  https://developer.mozilla.org/en-US/docs/Memory_Sanitizer).

**Comment by benesch** [comment]
Nice, :lgtm:! Though I feel pretty strongly that embedding that stack trace in the Makefile isn't necessary.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[Makefile, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcNc0_SlEmzQnPEvEF:-KjcNc0_SlEmzQnPEvEG:bo78wec) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L94)):*
> ```Makefile
> # #9  0x0000000000000001 in ?? ()
> # #10 0x00007fffffffee15 in ?? ()
> # #11 0x0000000000000000 in ?? ()
> ```

Is it really worth including this stack trace here? It makes our Makefile that much harder to digest and it's not like you'll be able to debug the issue from this one stack trace anyway.

---

*[Makefile, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcGm0nBnSwv_hAbdHD:-KjcGm0nBnSwv_hAbdHE:b-9ngrzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L119)):*
> ```Makefile
> override GOFLAGS += -installsuffix release-gnu
> override SUFFIX := $(SUFFIX)-linux-2.6.32-gnu-amd64
> BUILD_TYPE := release-gnu
> ```

Eep, are you sure changing this from `release` doesn't break anything?

---

*[build/common.mk, line 272 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcGdFiGX2eein9D7ZB:-KjcGdFiGX2eein9D7ZC:b-9ell0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/build/common.mk#L272)):*
> ```Makefile
> 
> # Go does not permit dashes in build tags. This is undocumented. Fun!
> NATIVE_SPECIFIER_TAG := $(subst -,_,$(NATIVE_SPECIFIER))$(STDMALLOC_SUFFIX)
> ```

Ah, whoops. Good catch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-:-KjcO0P-LF82SExPh31-:bt6pf1x)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[Makefile, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcNc0_SlEmzQnPEvEF:-KjcOsQmiKpPZBHVtVlG:b-tojrfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L94)):*
I'm not sure it makes it that much harder, but the upshot is that it's immediately more useful than the original version of the comment which read:

> # NB: using jemalloc with msan causes segfaults. Not sure why.

---

*[Makefile, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcGm0nBnSwv_hAbdHD:-KjcOnBpXdh8LdmfC-7Y:bp71ekr) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L119)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Eep, are you sure changing this from `release` doesn't break anything?
</blockquote></details>

Pretty sure; it's only used for display purposes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753)*

**Comment by benesch** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[Makefile, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcNc0_SlEmzQnPEvEF:-KjcVcaBDNwnOpbZFuQw:b-espj28) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L94)):*
Blergh. What if you link to a Gist instead? I think the comment is only marginally more useful. In fact I think your Reviewable comment

> I think we'd need the minimal repro to file it; for now I was discouraged by jemalloc/jemalloc#558 (comment) from producing that repro, but I may revisit.

> Interestingly, it doesn't look like anyone has tried using the two together, and even firefox's msan support turns off jemalloc (search for https://developer.mozilla.org/en-US/docs/Memory_Sanitizer in https://developer.mozilla.org/en-US/docs/Memory_Sanitizer).

is an order of magnitude more useful than the stack trace. I'll advocate for including a rewording of that comment instead of the stack trace.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753)*

**Comment by tamird** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[Makefile, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcNc0_SlEmzQnPEvEF:-KjceA0rDDlC59PcsK8R:bd1a6s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L94)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Blergh. What if you link to a Gist instead? I think the comment is only marginally more useful. In fact I think your Reviewable comment

> I think we'd need the minimal repro to file it; for now I was discouraged by jemalloc/jemalloc#558 (comment) from producing that repro, but I may revisit.

> Interestingly, it doesn't look like anyone has tried using the two together, and even firefox's msan support turns off jemalloc (search for https://developer.mozilla.org/en-US/docs/Memory_Sanitizer in https://developer.mozilla.org/en-US/docs/Memory_Sanitizer).

is an order of magnitude more useful than the stack trace. I'll advocate for including a rewording of that comment instead of the stack trace.
</blockquote></details>

Fiiiine. Filed upstream.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753)*

**Comment by tamird** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[Makefile, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-Kj_J35WobG1uiUYgCTZ-r1-64:-KjceED4vsTEypMl-80t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think we'd need the minimal repro to file it; for now I was discouraged by https://github.com/jemalloc/jemalloc/issues/558#issuecomment-272982746 from producing that repro, but I may revisit.

Interestingly, it doesn't look like anyone has tried using the two together, and even firefox's msan support turns off jemalloc (search for `https://developer.mozilla.org/en-US/docs/Memory_Sanitizer` in  https://developer.mozilla.org/en-US/docs/Memory_Sanitizer).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753)*

**Comment by benesch** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[Makefile, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753#-KjcNc0_SlEmzQnPEvEF:-Kjco8OiH7WaiLMDIinJ:bgc7vqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd69a546c5ed65acf4193bf6a80451b4ec9caf/Makefile#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Fiiiine. Filed upstream.
</blockquote></details>

:heart: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15753)*

**Pull request merged**